SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,181 @@
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME_CONDITION(USpatialPingComponent, ReplicatedPingID, COND_OwnerOnly);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	// Attempt to cast the owner of this component to the PlayerController class.
+	// Component will do nothing if cast fails.
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (OwningController == nullptr)
+	{
+		UE_LOG(LogSpatialPingComponent, Warning, TEXT("SpatialPingComponent did not find a valid owning PlayerController and will not function correctly. Ensure this component is only attached to a PlayerController."));
+	}
+
+	if (bStartWithPingEnabled)
+	{
+		SetPingEnabled(true);
+	}
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	// Clear up the timer before the component is disabled or destroyed.
+	SetPingEnabled(false);
+	Super::EndPlay(EndPlayReason);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	// Only execute if this component is attached to a player controller.
+	if (OwningController != nullptr)
+	{
+		// Only execute on owning local client.
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled)
+			{
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled)
+			{
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RoundTripPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	if (UWorld* World = GetWorld())
+	{
+		LastSentPingID = 0;
+		TimeoutCount = 0;
+		// Send a new ping, which will trigger a self-perpetuating sequence via timers.
+		SendNewPing();
+		// Set looping timer to 'tick' this component, it doesn't send any pings but matches the MinPingInterval for passing updates to the owning controller.
+		World->GetTimerManager().SetTimer(PingTickHandle, this, &USpatialPingComponent::TickPingComponent, MinPingInterval, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	if (UWorld* World = GetWorld())
+	{
+		bIsPingEnabled = false;
+		// Clear the timers.
+		World->GetTimerManager().ClearTimer(PingTickHandle);
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		// Reset ping output value.
+		RoundTripPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	// Pass latest measured ping to owning controller to be processed by PlayerState.
+	OwningController->UpdatePing(RoundTripPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	// Generate a new ping ID.
+	uint16 NewPingID = LastSentPingID + 1;
+	// Send and record new ping ID.
+	SendServerWorkerPingID(NewPingID);
+	LastSentPingID = NewPingID;
+	LastSentPingTimestamp = FPlatformTime::Seconds();
```

Comment:
Nit: should probably record the timestamp right before calling the RPC as opposed to after (though it would be surprising for one RPC to take more than 1ms).